File: src/main/java/xyz/phanta/cbtweaker/util/DummyBlockAccessor.java
package xyz.phanta.cbtweaker.util;

import net.minecraft.block.state.IBlockState;
import net.minecraft.init.Biomes;
import net.minecraft.init.Blocks;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.WorldType;
import net.minecraft.world.biome.Biome;

import javax.annotation.Nullable;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

public abstract class DummyBlockAccessor implements IBlockAccess {

    @Nullable
    @Override
    public TileEntity getTileEntity(BlockPos pos) {
        return null;
    }

    @Override
    public int getCombinedLight(BlockPos pos, int lightValue) {
        return lightValue;
    }

    @Override
    public boolean isAirBlock(BlockPos pos) {
        IBlockState state = getBlockState(pos);
        return state.getBlock().isAir(state, this, pos);
    }

    @Override
    public Biome getBiome(BlockPos pos) {
        return Biomes.PLAINS;
    }

    @Override
    public int getStrongPower(BlockPos pos, EnumFacing direction) {
        return 0;
    }

    @Override
    public WorldType getWorldType() {
        return WorldType.FLAT;
    }

    @Override
    public boolean isSideSolid(BlockPos pos, EnumFacing side, boolean _default) {
        return getBlockState(pos).isSideSolid(this, pos, side);
    }

    public static class MapBacked extends DummyBlockAccessor {

        private final Map<BlockPos, IBlockState> blockTable = new HashMap<>();

        @Override
        public IBlockState getBlockState(BlockPos pos) {
            IBlockState state = blockTable.get(pos);
            return state != null ? state : Blocks.AIR.getDefaultState();
        }

        public void setBlockState(BlockPos pos, IBlockState state) {
            blockTable.put(pos, state);
        }

        public Collection<Map.Entry<BlockPos, IBlockState>> getEntries() {
            return blockTable.entrySet();
        }

    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/util/world/DummyBlockAccessor.kt
package st.evening.mc.cbtweaker.util.world

import net.minecraft.block.state.IBlockState
import net.minecraft.init.Biomes
import net.minecraft.init.Blocks
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IBlockAccess
import net.minecraft.world.WorldType
import net.minecraft.world.biome.Biome
import st.evening.mc.prelude.api.util.game.ClientSide

abstract class DummyBlockAccessor : IBlockAccess {
    override fun getTileEntity(pos: BlockPos): TileEntity? = null

    @ClientSide.Physical
    override fun getCombinedLight(pos: BlockPos, lightValue: Int): Int = lightValue

    override fun isAirBlock(pos: BlockPos): Boolean = getBlockState(pos).let { it.block.isAir(it, this, pos) }

    @ClientSide.Physical
    override fun getBiome(pos: BlockPos): Biome = Biomes.PLAINS

    override fun getStrongPower(pos: BlockPos, direction: EnumFacing): Int = 0

    @ClientSide.Physical
    override fun getWorldType(): WorldType = WorldType.FLAT

    override fun isSideSolid(pos: BlockPos, side: EnumFacing, default: Boolean): Boolean =
        getBlockState(pos).isSideSolid(this, pos, side)

    class MapBacked : DummyBlockAccessor() {
        private val blockTable: MutableMap<BlockPos, IBlockState> = mutableMapOf()

        val entries: Collection<Map.Entry<BlockPos, IBlockState>>
            get() = blockTable.entries

        override fun getBlockState(pos: BlockPos): IBlockState = blockTable[pos] ?: Blocks.AIR.defaultState

        fun setBlockState(pos: BlockPos, state: IBlockState) {
            blockTable[pos] = state
        }
    }

}
